MidasFrontend/ArduinoHelper.cxx

```diff
@@ -0,0 +1,195 @@
+    if (dest_mm[AXIS_X] < gantry_x_min_mm || dest_mm[AXIS_X] > gantry_x_max_mm) {
+        cm_msg(
+            MERROR,
+            "validate_move_params",
+            "Destination on x-axis should be between %f mm and %f mm inclusive.",
+            gantry_x_min_mm, gantry_x_max_mm);
+        return false;
+    }
+
+    if (dest_mm[AXIS_Y] < gantry_y_min_mm || dest_mm[AXIS_Y] > gantry_y_max_mm) {
+        cm_msg(
+            MERROR,
+            "validate_move_params",
+            "Destination on y-axis should be between %f mm and %f mm inclusive.",
+            gantry_y_min_mm, gantry_y_max_mm);
+        return false;
+    }
+
+    if (vel_mm_s[AXIS_X] < gantry_vel_min_mm_s || vel_mm_s[AXIS_X] > gantry_vel_max_mm_s
+        || vel_mm_s[AXIS_Y] < gantry_vel_min_mm_s || vel_mm_s[AXIS_Y] > gantry_vel_max_mm_s) {
+        cm_msg(
+            MERROR,
+            "validate_move_params",
+            "Velocity should be between %f mm/s and %f mm/s inclusive.",
+            gantry_vel_min_mm_s, gantry_vel_max_mm_s);
+        return false;
+    }
+
+    return true;
+}
+
+static int32_t mm_to_cts(float val_mm) {
+    return round(val_mm / mm_cts_ratio);
+}
+
+static uint32_t mm_to_steps(float val_mm) {
+    return round(val_mm / mm_steps_ratio);
+}
+
+static AxisDirection get_direction(int32_t displacement) {
+    return (displacement < 0 ? AXIS_DIR_NEGATIVE : AXIS_DIR_POSITIVE);
+}
+
+static bool handle_serial_result(SerialResult res) {
+    if (res == SERIAL_OK) return true;
+
+    printf("Serial Error: %d\n", res);
+    return false;
+}
+
+static bool handle_axis_result(AxisResult res) {
+    if (res == AXIS_OK) return true;
+
+    printf("Axis Error: %d\n", res);
+    return false;
+}
+
+static bool attempt_move_axis(AxisId axis, int32_t cur_pos_counts, float dest_mm, float vel_mm_s)
+{
+    // Target position
+    int32_t target_counts = mm_to_cts(dest_mm);
+
+    // Relative displacement
+    int32_t disp_counts = (target_counts - cur_pos_counts);
+
+    // Direction
+    AxisDirection dir = get_direction(disp_counts);
+
+    // Velocity
+    uint32_t vel_steps_s = mm_to_steps(vel_mm_s);
+
+    // Send command
+    SerialResult ser_res;
+    AxisResult axis_res;
+    ser_res = comm.move(axis, dir, vel_steps_s, abs(disp_counts), &axis_res, MSG_RECEIVE_TIMEOUT);
+
+    // Handle results
+    return (handle_serial_result(ser_res) && handle_axis_result(axis_res));
+}
+
+/*****************************************************************************/
+/*                             PUBLIC FUNCTIONS                              */
+/*****************************************************************************/
+
+/**
+ * @brief Establishes a serial connection to the Arduino
+ * 
+ * This includes opening the serial device as well as waiting to receive a
+ * ping message from the Arduino to validate that it is running
+ * 
+ * @return true if the connection was successfully established, otherwise false
+ */
+bool arduino_connect(char *device_file)
```

Comment:
is timeout defined in device file? or does this not timeout?

Reply:
There is no timeout for this. This is the first thing that happens when you start the program so the idea is to just sit there and wait until the Arduino pings us.